primitives/src/identity_id.rs

```diff
@@ -165,6 +166,84 @@ impl Printable for IdentityId {
     }
 }
 
+/// A wrapper for a portfolio name. It is used for non-default (aka "user") portfolios only since
+/// default ones are nameless.
+#[derive(
+    Decode, Encode, Clone, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, VecU8StrongTyped,
+)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct PortfolioName(pub Vec<u8>);
+
+/// The unique ID of a non-default portfolio.
+pub type PortfolioNumber = u64;
+
+#[derive(Decode, Encode, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum PortfolioKind {
+    /// The default portfolio of a DID.
+    Default,
+    /// A user-defined portfolio of a DID.
+    User(PortfolioNumber),
+}
+
+impl Default for PortfolioKind {
+    fn default() -> Self {
+        PortfolioKind::Default
+    }
+}
+
+/// The ID of a portfolio.
+#[derive(Decode, Encode, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct PortfolioId {
+    /// The DID of the portfolio.
+    did: IdentityId,
+    /// The kind of the portfolio: either default or user.
+    kind: PortfolioKind,
+}
+
+impl Default for PortfolioId {
+    fn default() -> Self {
+        PortfolioId {
+            did: IdentityId::default(),
+            ..Default::default()
+        }
+    }
+}
+
+impl Printable for PortfolioId {
+    fn print(&self) {
+        self.did.print();
+        sp_io::misc::print_utf8(b"/");
+        match self.kind {
+            PortfolioKind::Default => {
+                sp_io::misc::print_utf8(b"default");
+            }
+            PortfolioKind::User(num) => {
+                sp_io::misc::print_hex(&num.to_be_bytes());
+            }
+        }
+    }
+}
+
+impl PortfolioId {
+    /// Returns the default portfolio of `did`.
+    pub fn default_portfolio(did: IdentityId) -> PortfolioId {
+        PortfolioId {
+            did,
+            kind: PortfolioKind::Default,
+        }
+    }
```

Comment:
Nit:
```suggestion
    pub fn default_portfolio(did: IdentityId) -> Self {
        Self {
            did,
            kind: PortfolioKind::Default,
        }
    }
```

Reply:
Is this the best practice or just personal choice?

Reply:
Imo best practice as it reduces the number of places you might need to make future changes & makes it obvious that you're referring to the current type, but it's just a nit. :)

Reply:
+1 on using `Self`. I've been slowly refactoring our codebase to use `Self` whenever I touch any of these.